File: Difficulty: Medium/Evaluation of Postfix Expression/evaluation-of-postfix-expression.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends

class Solution {
  public:
    int evaluate(vector<string>& arr) {
        int n=arr.size();
        stack<int>st;
        for(int i=0;i<n;i++){
            if(arr[i]!="*" && arr[i]!="/" && arr[i]!="-" && arr[i]!="+"){
                string s=arr[i];
                int x=s.length();
                int num=0;
                int flag=0;
                int i=0;
                if(s[i]=='-'){
                    flag=1;
                    i++;
                    x--;
                }
                while(x--){
                    num=num*10+(s[i]-'0');
                    i++;
                }
                if(flag==1) num=-1*num;
                st.push(num);
            }else{
                int b=st.top();
                st.pop();
                int a=st.top();
                st.pop();
                int c;
                if(arr[i]=="*") c=a*b;
                else if(arr[i]=="/") c=a/b;
                else if(arr[i]=="-") c=a-b;
                else if(arr[i]=="+") c=a+b;
                st.push(c);
            }
        }
        return st.top();
    }
};


//{ Driver Code Starts.

int main() {
    string str;
    getline(cin, str);
    int t = stoi(str);
    while (t--) {
        getline(cin, str);
        stringstream ss(str);
        vector<string> arr;
        while (ss >> str) {
            arr.push_back(str);
        }
        Solution ob;
        cout << ob.evaluate(arr) << endl;
        cout << "~" << endl;
    }
}
// } Driver Code Ends

File: Difficulty: Medium/Evaluation of Postfix Expression/evaluation-of-postfix-expression.java
//{ Driver Code Starts
// Initial Template for Java
import java.io.*;
import java.util.*;


// } Driver Code Ends


class Solution {
    Stack<Integer> stack = new Stack<>();
    public int evaluate(String[] arr) {
        for(String token: arr){
            if(isOperator(token)){
                int b = stack.pop();
                int a = stack.pop();
                stack.push(calculation(a,b,token));
            }
            else
            {
                stack.push(Integer.parseInt(token));
            }
        }
        
        return stack.pop();
        
    }
    
    public static boolean isOperator(String token){
        return token.equals("+")|| token.equals("-") || token.equals("*") || token.equals("/");
    }
    
    
    public static int calculation(int a , int b , String operator){
        switch(operator){
            case "+" : return a + b ;
            case "-" : return a- b;
            case "*" : return a * b;
            case "/" : return a / b;
            default: throw new IllegalArgumentException("Invalid operator");
        }
    }
}


//{ Driver Code Starts.

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        String line = reader.readLine();
        int t = Integer.parseInt(line);
        while (t-- > 0) {
            line = reader.readLine();
            String[] arr = line.split(" ");
            Solution solution = new Solution();
            System.out.println(solution.evaluate(arr));
            System.out.println("~");
        }
    }
}

// } Driver Code Ends